csp_billing_adapter/adapter.py

```diff
@@ -141,60 +143,63 @@ def event_loop_handler(
-                'errors': [
-                    f'Usage data retrieval failed: {exc}'
-                ]
-            },
-            replace=False
-        )
-        # NOTE:
-        # If this happens during initial deployment timeframe, failing
-        # to retrieve usage data here could simple mean that there is
-        # nothing to bill for yet so not proceeding to metering makes
-        # sense.
-        # If this occurs for an existing deployment, with valid usage
-        # records waiting to be billed, continuing on to metering is
-        # worth considering.
-        # However, doing so will currently result in the error state
-        # that was saved in the csp_config being lost, as it would be
-        # overwritten with a success state, or a different error.
+        errors.append(f'Usage data retrieval failed: {exc}')
 
-        log.info('Finishing event loop processing - get_usage_data() error')
-        return now
-
-    log.debug('Retrieved usage data: %s', usage)
-
-    add_usage_record(hook, config, usage)
+    if usage:
+        try:
+            add_usage_record(hook, config, usage)
+        except Exception as error:
+            log.warning("Failed to save usage data to cache: %s", str(error))
+            errors.append(f'Usage data failed to save: {error}')
 
     # handle metering/billing updates
-    cache = hook.get_cache(config=config)
-    now = get_now()
-
-    log.debug(
-        "Now: %s, Next Reporting Time: %s, Next Bill Time: %s",
-        date_to_string(now),
-        cache['next_reporting_time'],
-        cache['next_bill_time']
-    )
+    try:
+        cache = hook.get_cache(config=config)
+    except Exception as error:
+        log.warning("Failed to retrieve cache: %s", str(error))
+        errors.append(f'Cache retrieval failed: {error}')
+        cache = {}
 
-    if now >= string_to_date(cache['next_bill_time']):
-        log.info('Attempt a billing cycle update')
-        process_metering(config, cache, hook)
-    elif now >= string_to_date(cache['next_reporting_time']):
-        log.info('Attempt a reporting cycle update')
-        process_metering(config, cache, hook, empty_metering=True)
+    if cache:
+        log.debug(
+            "Now: %s, Next Reporting Time: %s, Next Bill Time: %s",
+            date_to_string(now),
+            cache['next_reporting_time'],
+            cache['next_bill_time']
+        )
 
+        if now >= string_to_date(cache['next_bill_time']):
+            log.info('Attempt a billing cycle update')
+            errors += process_metering(
+                config,
+                cache,
+                hook
+            )
+        elif now >= string_to_date(cache['next_reporting_time']):
+            log.info('Attempt a reporting cycle update')
+            errors += process_metering(
+                config,
+                cache,
+                hook,
+                empty_metering=True
+            )
+
+    hook.update_csp_config(
```

Comment:
Since we have caught and continued from all exceptions encountered, none are treated as hard failures, so we would always execute this block, right? And come back again on the next loop, and so on...

Just wondering if we should perhaps be setting a flag in the earlier exception catching blocks to indicate a hard failure scenario, and then, if the flag is set, raise a hard failure exception after updating the csp_config?

As discussed elsewhere, we could structure our exceptions such that they are derived from 3 base exception classes, e.g. hard/fail errors, soft/finish event loop errors, and soft/continue event loop errors. Then when we catch an exception we could check if the exception is derived from the hard/fail error type, and if so set the flag variable to indicate that the event loop should hard fail...

Reply:
Certain cases are hard failures and not caught such as inside process billing. Those we probably want to catch and raise with more info. But for now they would cause the process to exit.

Reply:
It might help, and of course I was not part of the other conversation about hard/soft failures and thus can only opinion on this snippet, to thin about the meaning of this for the customer and us. If/when we cannot add a usage record we will not bill the customer correctly, the calculation of an average for NV for example could be skewed in either our or the customers favor. For example in hour 1 the customer manages 1000 nodes, then the rest of the month only 5 nodes get managed. If now we have a failure to add usage records for the rest of the month the customer gets billed for managing 1000 nodes, problem we over bill big time. If on the other hand we have a failure adding the usage record in hour 1 but then the rest of the month there is no issue we'd bill the customer for 5 nodes as that would be the minimum. So the question is how do we bubble this up to the customer such that they call support and get the issue resolved?

Reply:
Yeah, but that may get convoluted in this proposed workflow...

If process_metering() hits an hard error scenario, and re-raised it, it wouldn't have a complete picture of the overall error state at that point so we would need to catch the error again in event_loop_handler(), and add the error the error list as well as setting a flag indicating we need to raise a hard failure, then fall through to updating the csp_config with the accumulated errors, and finally raise the hard error at that point, without any extra exception context to inherit...

Alternatively we could pass the existing errors list as an argument to process_metering(), and if it hit a hard error scenario it could include the error list in it's csp_config update, and then re-raise with our hard error exception, without having to catch any exception in the event_loop_handler()... But we would need to ensure that that the process_metering() exception handling csp_config update also sets the timestamp etc fields appropriately...

Reply:
Adapter only can log errors and hand them out via "API". It's up to the app UI to get these to the user. I don't think there's anything to do in that regard in adapter.

Reply:
Yes, the the hard failure scenarios need more work. However, this may only be possible as logs. For example if config map cannot be updated for some reason there's nothing adapter can do except try to log the current state and issues. Ditto for cache. However, dumping cache to log would be a challenge when it has ~700 records.